File: langchain-module/src/main/python/reviewpilot_langchain/agents/openai_agent.py
"""
OpenAI agent implementation using LangChain
"""

import time
from typing import List, Dict, Any
from langchain_openai import ChatOpenAI
from langchain.agents import initialize_agent, AgentType, Tool
from langchain.memory import ConversationBufferMemory
from langchain.prompts import PromptTemplate
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain.vectorstores import Chroma
from langchain.embeddings import OpenAIEmbeddings
from langchain.schema import Document
from langchain.chains import LLMChain
from .base_agent import BaseAgent
from ..models import PullRequestDetails, CodeReviewResult, ReviewConfig


class OpenAIAgent(BaseAgent):
    """OpenAI agent implementation using LangChain"""
    
    def _setup_llm(self):
        """Setup OpenAI language model"""
        self.llm = ChatOpenAI(
            model_name=self.config.model_name,
            temperature=self.config.temperature,
            max_tokens=self.config.max_tokens
        )
    
    def _setup_tools(self):
        """Setup tools for code analysis"""
        self.tools = [
            Tool(
                name="code_analyzer",
                func=self._analyze_code_snippet,
                description="Analyze a specific code snippet for issues and improvements"
            ),
            Tool(
                name="security_checker",
                func=self._check_security_issues,
                description="Check for common security vulnerabilities in code"
            ),
            Tool(
                name="performance_analyzer",
                func=self._analyze_performance,
                description="Analyze code for performance issues and optimization opportunities"
            )
        ]
    
    def _setup_memory(self):
        """Setup conversation memory"""
        if self.config.enable_memory:
            self.memory = ConversationBufferMemory(
                memory_key="chat_history",
                return_messages=True
            )
        else:
            self.memory = None
    
    def review_pull_request(self, pr_details: PullRequestDetails) -> CodeReviewResult:
        """Review a pull request using LangChain agent"""
        start_time = time.time()
        
        # Initialize agent
        agent = initialize_agent(
            tools=self.tools,
            llm=self.llm,
            agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION,
            memory=self.memory,
            verbose=True
        )
        
        # Create review prompt
        review_prompt = self._create_review_prompt(pr_details)
        
        # Run the review
        response = agent.run(review_prompt)
        
        # Analyze different aspects
        security_issues = self.analyze_security(pr_details)
        performance_issues = self.analyze_performance(pr_details)
        quality_metrics = self.analyze_code_quality(pr_details)
        
        # Calculate scores
        quality_score = self._calculate_quality_score(quality_metrics)
        confidence_score = self._calculate_confidence_score(response)
        
        review_time = time.time() - start_time
        
        return CodeReviewResult(
            summary=response,
            issues=quality_metrics.get("issues", []),
            suggestions=quality_metrics.get("suggestions", []),
            security_concerns=security_issues,
            performance_issues=performance_issues,
            code_quality_score=quality_score,
            confidence_score=confidence_score,
            review_time=review_time,
            tokens_used=self._estimate_tokens(response),
            model_used=self.config.model_name
        )
    
    def analyze_security(self, pr_details: PullRequestDetails) -> List[str]:
        """Analyze security aspects of the code"""
        security_prompt = f"""
        Analyze the following code changes for security vulnerabilities:
        
        Title: {pr_details.title}
        Description: {pr_details.description}
        Changed Files: {pr_details.changed_files}
        Diff: {pr_details.diff}
        
        Focus on:
        1. SQL injection vulnerabilities
        2. XSS vulnerabilities
        3. Authentication/authorization issues
        4. Input validation problems
        5. Sensitive data exposure
        6. Insecure dependencies
        
        Provide specific security concerns found:
        """
        
        response = self.llm.predict(security_prompt)
        return [line.strip() for line in response.split('\n') if line.strip()]
    
    def analyze_performance(self, pr_details: PullRequestDetails) -> List[str]:
        """Analyze performance aspects of the code"""
        performance_prompt = f"""
        Analyze the following code changes for performance issues:
        
        Title: {pr_details.title}
        Description: {pr_details.description}
        Changed Files: {pr_details.changed_files}
        Diff: {pr_details.diff}
        
        Focus on:
        1. Algorithm efficiency
        2. Database query optimization
        3. Memory usage
        4. Network calls optimization
        5. Caching opportunities
        6. Resource leaks
        
        Provide specific performance issues found:
        """
        
        response = self.llm.predict(performance_prompt)
        return [line.strip() for line in response.split('\n') if line.strip()]
    
    def analyze_code_quality(self, pr_details: PullRequestDetails) -> Dict[str, Any]:
        """Analyze code quality and return metrics"""
        quality_prompt = f"""
        Analyze the following code changes for code quality:
        
        Title: {pr_details.title}
        Description: {pr_details.description}
        Changed Files: {pr_details.changed_files}
        Diff: {pr_details.diff}
        
        Provide analysis in the following format:
        
        ISSUES:
        - [List specific issues found]
        
        SUGGESTIONS:
        - [List improvement suggestions]
        
        COMPLEXITY: [Low/Medium/High]
        READABILITY: [Low/Medium/High]
        MAINTAINABILITY: [Low/Medium/High]
        """
        
        response = self.llm.predict(quality_prompt)
        
        # Parse response
        issues = []
        suggestions = []
        complexity = "Medium"
        readability = "Medium"
        maintainability = "Medium"
        
        lines = response.split('\n')
        current_section = None
        
        for line in lines:
            line = line.strip()
            if line.startswith('ISSUES:'):
                current_section = 'issues'
            elif line.startswith('SUGGESTIONS:'):
                current_section = 'suggestions'
            elif line.startswith('COMPLEXITY:'):
                complexity = line.split(':')[1].strip()
            elif line.startswith('READABILITY:'):
                readability = line.split(':')[1].strip()
            elif line.startswith('MAINTAINABILITY:'):
                maintainability = line.split(':')[1].strip()
            elif line.startswith('-') and current_section:
                item = line[1:].strip()
                if current_section == 'issues':
                    issues.append(item)
                elif current_section == 'suggestions':
                    suggestions.append(item)
        
        return {
            "issues": issues,
            "suggestions": suggestions,
            "complexity": complexity,
            "readability": readability,
            "maintainability": maintainability
        }
    
    def _create_review_prompt(self, pr_details: PullRequestDetails) -> str:
        """Create a comprehensive review prompt"""
        if self.config.custom_prompt:
            prompt = self.config.custom_prompt
        else:
            prompt = """
            You are an expert code reviewer. Analyze the following pull request and provide a comprehensive review.
            
            Pull Request Details:
            - Title: {title}
            - Description: {description}
            - Changed Files: {changed_files}
            - Diff: {diff}
            
            Please provide:
            1. Overall assessment of the changes
            2. Code quality analysis
            3. Potential issues and concerns
            4. Suggestions for improvement
            5. Security considerations
            6. Performance implications
            
            Be specific and actionable in your feedback.
            """
        
        return prompt.format(
            title=pr_details.title,
            description=pr_details.description,
            changed_files=pr_details.changed_files,
            diff=pr_details.diff
        )
    
    def _analyze_code_snippet(self, code: str) -> str:
        """Analyze a specific code snippet"""
        prompt = f"Analyze this code snippet for issues and improvements:\n\n{code}"
        return self.llm.predict(prompt)
    
    def _check_security_issues(self, code: str) -> str:
        """Check for security issues in code"""
        prompt = f"Check this code for security vulnerabilities:\n\n{code}"
        return self.llm.predict(prompt)
    
    def _analyze_performance(self, code: str) -> str:
        """Analyze performance aspects of code"""
        prompt = f"Analyze this code for performance issues:\n\n{code}"
        return self.llm.predict(prompt)
    
    def _calculate_quality_score(self, metrics: Dict[str, Any]) -> float:
        """Calculate code quality score"""
        scores = {
            "Low": 30,
            "Medium": 60,
            "High": 90
        }
        
        complexity_score = scores.get(metrics.get("complexity", "Medium"), 60)
        readability_score = scores.get(metrics.get("readability", "Medium"), 60)
        maintainability_score = scores.get(metrics.get("maintainability", "Medium"), 60)
        
        return (complexity_score + readability_score + maintainability_score) / 3
    
    def _calculate_confidence_score(self, response: str) -> float:
        """Calculate confidence score based on response quality"""
        # Simple heuristic based on response length and structure
        if len(response) < 100:
            return 30.0
        elif len(response) < 500:
            return 60.0
        else:
            return 90.0
    
    def _estimate_tokens(self, text: str) -> int:
        """Estimate token count"""
        # Rough estimation: 1 token ≈ 4 characters
        return len(text) // 4 

File: langchain4j-module/src/main/java/com/csharma/reviewpilot/langchain4j/agent/OpenAIAgent.java
package com.csharma.reviewpilot.langchain4j.agent;

import com.csharma.reviewpilot.langchain4j.model.CodeReviewResult;
import com.csharma.reviewpilot.langchain4j.model.PullRequestDetails;
import com.csharma.reviewpilot.langchain4j.model.ReviewConfig;
import dev.langchain4j.data.message.AiMessage;
import dev.langchain4j.data.message.ChatMessage;
import dev.langchain4j.data.message.SystemMessage;
import dev.langchain4j.data.message.UserMessage;
// Memory functionality will be added later when dependencies are available
import dev.langchain4j.model.chat.ChatLanguageModel;
import dev.langchain4j.model.openai.OpenAiChatModel;
import dev.langchain4j.model.output.Response;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Duration;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * OpenAI agent implementation using LangChain4j
 */
public class OpenAIAgent implements CodeReviewAgent {
    private static final Logger logger = LoggerFactory.getLogger(OpenAIAgent.class);
    
    private final ReviewConfig config;
    private final ChatLanguageModel chatModel;
    // private final ChatMemory chatMemory; // Will be enabled when memory dependency is available

    public OpenAIAgent(ReviewConfig config) {
        this.config = config;
        
        // Initialize OpenAI chat model
        this.chatModel = OpenAiChatModel.builder()
                .apiKey(System.getenv("OPENAI_API_KEY"))
                .modelName(config.getModelName())
                .temperature(config.getTemperature())
                .maxTokens(config.getMaxTokens())
                .timeout(Duration.ofSeconds(60))
                .build();
        
        // Initialize chat memory if enabled
        // if (config.getEnableMemory()) {
        //     this.chatMemory = MessageWindowChatMemory.withMaxMessages(10);
        // } else {
        //     this.chatMemory = null;
        // }
        
        logger.info("Initialized OpenAI agent with model: {}", config.getModelName());
    }

    @Override
    public CodeReviewResult reviewPullRequest(PullRequestDetails prDetails) {
        long startTime = System.currentTimeMillis();
        logger.info("Starting code review for PR: {}", prDetails.getTitle());
        
        try {
            // Create review prompt
            String reviewPrompt = createReviewPrompt(prDetails);
            
            // Prepare messages
            List<ChatMessage> messages = new ArrayList<>();
            
            // Add system message
            messages.add(new SystemMessage("You are an expert code reviewer. Provide comprehensive, actionable feedback."));
            
            // Add user message with review prompt
            messages.add(new UserMessage(reviewPrompt));
            
            // Generate review
            Response<AiMessage> response = chatModel.generate(messages);
            String reviewSummary = response.content().text();
            
            // Calculate review time
            long reviewTime = System.currentTimeMillis() - startTime;
            
            // Create result
            CodeReviewResult result = new CodeReviewResult(reviewSummary);
            result.setReviewTime((double) reviewTime / 1000.0); // Convert to seconds
            result.setModelUsed(config.getModelName());
            result.setTokensUsed(estimateTokens(reviewSummary));
            result.setConfidenceScore(calculateConfidenceScore(reviewSummary));
            
            // Perform additional analyses if enabled
            if (config.getIncludeSecurityAnalysis()) {
                result.setSecurityConcerns(analyzeSecurity(prDetails));
            }
            
            if (config.getIncludePerformanceAnalysis()) {
                result.setPerformanceIssues(analyzePerformance(prDetails));
            }
            
            if (config.getIncludeCodeQualityAnalysis()) {
                result.setIssues(analyzeCodeQuality(prDetails));
                result.setSuggestions(generateSuggestions(prDetails));
                result.setCodeQualityScore(calculateQualityScore(prDetails));
            }
            
            logger.info("Completed code review in {} seconds", result.getReviewTime());
            return result;
            
        } catch (Exception e) {
            logger.error("Error during code review: {}", e.getMessage());
            throw new RuntimeException("Failed to review pull request: " + e.getMessage(), e);
        }
    }

    private String createReviewPrompt(PullRequestDetails prDetails) {
        if (config.getCustomPrompt() != null && !config.getCustomPrompt().isEmpty()) {
            return config.getCustomPrompt()
                    .replace("{{title}}", prDetails.getTitle())
                    .replace("{{description}}", prDetails.getDescription())
                    .replace("{{changedFiles}}", prDetails.getChangedFiles().toString())
                    .replace("{{diff}}", prDetails.getDiff());
        }
        
        return String.format("""
            Please review the following pull request and provide comprehensive feedback:
            
            Title: %s
            Description: %s
            Changed Files: %s
            Diff: %s
            
            Please provide:
            1. Overall assessment of the changes
            2. Code quality analysis
            3. Potential issues and concerns
            4. Suggestions for improvement
            5. Security considerations
            6. Performance implications
            
            Be specific and actionable in your feedback.
            """, 
            prDetails.getTitle(),
            prDetails.getDescription(),
            prDetails.getChangedFiles(),
            prDetails.getDiff()
        );
    }

    private List<String> analyzeSecurity(PullRequestDetails prDetails) {
        String securityPrompt = String.format("""
            Analyze the following code changes for security vulnerabilities:
            
            Title: %s
            Description: %s
            Changed Files: %s
            Diff: %s
            
            Focus on:
            1. SQL injection vulnerabilities
            2. XSS vulnerabilities
            3. Authentication/authorization issues
            4. Input validation problems
            5. Sensitive data exposure
            6. Insecure dependencies
            
            Provide specific security concerns found:
            """, 
            prDetails.getTitle(),
            prDetails.getDescription(),
            prDetails.getChangedFiles(),
            prDetails.getDiff()
        );
        
        try {
            Response<AiMessage> response = chatModel.generate(new UserMessage(securityPrompt));
            String result = response.content().text();
            return parseListFromResponse(result);
        } catch (Exception e) {
            logger.warn("Error analyzing security: {}", e.getMessage());
            return new ArrayList<>();
        }
    }

    private List<String> analyzePerformance(PullRequestDetails prDetails) {
        String performancePrompt = String.format("""
            Analyze the following code changes for performance issues:
            
            Title: %s
            Description: %s
            Changed Files: %s
            Diff: %s
            
            Focus on:
            1. Algorithm efficiency
            2. Database query optimization
            3. Memory usage
            4. Network calls optimization
            5. Caching opportunities
            6. Resource leaks
            
            Provide specific performance issues found:
            """, 
            prDetails.getTitle(),
            prDetails.getDescription(),
            prDetails.getChangedFiles(),
            prDetails.getDiff()
        );
        
        try {
            Response<AiMessage> response = chatModel.generate(new UserMessage(performancePrompt));
            String result = response.content().text();
            return parseListFromResponse(result);
        } catch (Exception e) {
            logger.warn("Error analyzing performance: {}", e.getMessage());
            return new ArrayList<>();
        }
    }

    private List<Map<String, Object>> analyzeCodeQuality(PullRequestDetails prDetails) {
        // Simplified implementation - in a real scenario, you'd do more detailed analysis
        List<Map<String, Object>> issues = new ArrayList<>();
        
        // Example issues
        if (prDetails.getChangedFiles().size() > 10) {
            Map<String, Object> issue = new HashMap<>();
            issue.put("type", "complexity");
            issue.put("message", "Too many files changed in single PR");
            issue.put("severity", "medium");
            issues.add(issue);
        }
        
        return issues;
    }

    private List<String> generateSuggestions(PullRequestDetails prDetails) {
        // Simplified implementation
        List<String> suggestions = new ArrayList<>();
        suggestions.add("Consider adding unit tests for the changes");
        suggestions.add("Review the code for potential edge cases");
        return suggestions;
    }

    private Double calculateQualityScore(PullRequestDetails prDetails) {
        // Simplified scoring algorithm
        double score = 80.0; // Base score
        
        // Adjust based on number of files changed
        if (prDetails.getChangedFiles().size() > 10) {
            score -= 10;
        }
        
        // Adjust based on description quality
        if (prDetails.getDescription().length() < 50) {
            score -= 5;
        }
        
        return Math.max(0.0, Math.min(100.0, score));
    }

    private Double calculateConfidenceScore(String reviewSummary) {
        // Simple heuristic based on response length and structure
        if (reviewSummary.length() < 100) {
            return 30.0;
        } else if (reviewSummary.length() < 500) {
            return 60.0;
        } else {
            return 90.0;
        }
    }

    private Integer estimateTokens(String text) {
        // Rough estimation: 1 token ≈ 4 characters
        return text.length() / 4;
    }

    private List<String> parseListFromResponse(String response) {
        List<String> items = new ArrayList<>();
        String[] lines = response.split("\n");
        
        for (String line : lines) {
            line = line.trim();
            if (line.startsWith("-") || line.startsWith("•") || line.startsWith("*")) {
                items.add(line.substring(1).trim());
            } else if (line.matches("\\d+\\..*")) {
                items.add(line.replaceFirst("\\d+\\.", "").trim());
            }
        }
        
        return items;
    }

    @Override
    public String getAgentName() {
        return "openai";
    }
} 